lfs/http.go

```diff
@@ -1,30 +1,112 @@
 package lfs
 
 import (
+	"bufio"
+	"bytes"
 	"crypto/tls"
 	"errors"
 	"fmt"
 	"io"
 	"net"
 	"net/http"
+	"net/http/httputil"
 	"os"
+	"path/filepath"
 	"strings"
+	"sync"
 	"time"
 
 	"github.com/github/git-lfs/vendor/_nuts/github.com/rubyist/tracerx"
 )
 
-func DoHTTP(c *Configuration, req *http.Request) (*http.Response, error) {
-	traceHttpRequest(c, req)
-	res, err := c.HttpClient().Do(req)
+type transferStats struct {
+	HeaderSize int
+	BodySize   int
+	Start      time.Time
+	Stop       time.Time
+}
+
+type transfer struct {
+	requestStats  *transferStats
+	responseStats *transferStats
+	url           string
+}
+
+var (
+	// TODO should use some locks
+	transfers           = make(map[*http.Response]*transfer)
+	transferBuckets     = make(map[string][]*http.Response)
+	transfersLock       sync.Mutex
+	transferBucketsLock sync.Mutex
+)
+
+func LogTransfer(key string, res *http.Response) {
+	if Config.isLoggingStats {
+		transferBucketsLock.Lock()
+		transferBuckets[key] = append(transferBuckets[key], res)
+		transferBucketsLock.Unlock()
+	}
+}
+
+type HttpClient struct {
+	*http.Client
+}
+
+func (c *HttpClient) Do(req *http.Request) (*http.Response, error) {
+	traceHttpRequest(req)
+
+	crc := countingRequest(req)
+	if req.Body != nil {
+		// Only set the body if we have a body, but create the countingRequest
+		// anyway to make using zeroed stats easier.
+		req.Body = crc
+	}
+
+	start := time.Now()
+	res, err := c.Client.Do(req)
+	if err != nil {
+		return res, err
+	}
+
+	traceHttpResponse(res)
+
+	cresp := countingResponse(res)
+	res.Body = cresp
+
+	if Config.isLoggingStats {
+		reqHeaderSize := 0
+		resHeaderSize := 0
+
+		if dump, err := httputil.DumpRequest(req, false); err == nil {
+			reqHeaderSize = len(dump)
+		}
+
+		if dump, err := httputil.DumpResponse(res, false); err == nil {
+			resHeaderSize = len(dump)
+		}
+
+		reqstats := &transferStats{HeaderSize: reqHeaderSize, BodySize: crc.Count}
+
+		// Response body size cannot be figured until it is read. Do not rely on a Content-Length
+		// header because it may not exist or be -1 in the case of chunked responses.
+		resstats := &transferStats{HeaderSize: resHeaderSize, Start: start}
+		transfersLock.Lock()
+		transfers[res] = &transfer{requestStats: reqstats, responseStats: resstats, url: req.URL.String()}
```

Comment:
It may not be necessary to dump the url, since you can get it from the response object with `res.Request.URL`.

Maybe it makes too little difference, but you could build the `transfer` object outside of the lock.

```golang
t := &transfer{requestStats: reqstats, responseStats: resstats, url: req.URL.String()}
transfersLock.Lock()
transfers[res] = t
transfersLock.Unlock()
```